### PR TITLE
Documentation for on demand scan using APIs

### PR DESCRIPTION
**Pull request by Aditi-sha**

### CODE REVIEW

File: 07-On demand scan as an API/README.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+$header = "Bearer " + $token.AccessToken
+$headers = @{"Authorization"=$header;"Content-Type"="application/json";}
+$requestBody = @{"SubscriptionIDList"=@("subscription_id");}
+
+# Example:
+$requestBody = @{"SubscriptionIDList"=@("sub1","sub2");}
+
+
+$method = [Microsoft.PowerShell.Commands.WebRequestMethod]::POST
+$apiUri = "https://azsk-ats-api-ringn.azurewebsites.net/adhocscan/RequestScan"
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md]: Please correct URL: https://AzSK-AzTS-WebApi-xxxxx.azurewebsites.net/adhocscan/RequestScan".

File: 07-On demand scan as an API/README.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+$token = Get-MsalToken -TenantId '<tenant-id>' -ClientId '<client-app-id>' -RedirectUri 'https://localhost' -Scopes '<WebAPI-scope>'
+
+```
+
+[Back to top…](README.md#On-this-page)
+
+## API to scan a subscription
+
+Use below PowerShell command to get control scan result.
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:115]: You should add a description of the API.
Also, the data contract explaining various scenarios of this API.

File: 07-On demand scan as an API/README.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+|----|----|----|
+| SubscriptionId| List of subscription id(s) for scan. | Yes|
+
+</br>
+
+[Back to top…](README.md#On-this-page)
+
+## API to get control scan result
+Use below PowerShell command to get control scan result.
+
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:148]: Same as above. Description + Data contract

File: 07-On demand scan as an API/README.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+| ScanRequestId | To get control scan result with respect to the scan request id.| Yes |
+| ControlIdList| List of control ids to get control result only for specific controls.| No |
+| ResourceNameList | List of resources to get control result only for certain resources.| No |
+
+
+</br>
+
+> **Note:** 
+> </br>
+> _1. Enable the flag 'OnDemandScanAPI' in order to expose API endpoints._
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:195]: This is a note for admin. This should have a separate section.

File: 07-On demand scan as an API/README.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+> </br>
+> _2. On demand scan for each subcription can be requested maximum 10 times in a day._
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:197]: We should specify this after the description.

File: 07-On demand scan as an API/README.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+> </br>
+> _3. You can get entire control result only for one subscription at a time._
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:199]: This is API specific.

File: 07-On demand scan as an API/README.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+> </br>
+> _4. If ScanRequestId is empty then API will return latest control scan result._
+> </br>
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:202]: You can mention #4 as part of **Request body parameter details:**

File: 07-On demand scan as an API/README.md [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+### Using user authentication code flow
+User authentication code flow uses user's crediential to generate the token. User must have access over the subscription  to scan or to get the control scan result.
+
+<!-- TBD: Also, 'Admin consent' is required only when the access of granting consent is restricted to admin at AD level -->
+
+**Note for WebAPI Owner:**
+> WebAPI must have 'Admin consent' granted to expose the APIs.
+> </br>
+> Grant admin consent for client app registration:
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:121]: Add step to get client id from Web API config: AADClientAppDetails__ClientId

File: 07-On demand scan as an API/README.md [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+> 1. Go to Azure Portal.
+> 2. Go to 'App Registration'.
+> 3. Select appropriate 'App Registration'.
+> 4. Go to 'API Permissions'.
+> 5. Click 'Grant admin consent for Microsoft'.
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:126]: Grant admin consent for your tenant

Comment (SINIKI) [inline 07-On demand scan as an API/README.md:126]: You should mention the scope for which consent has to be granted.

Comment (SINIKI) [inline 07-On demand scan as an API/README.md:126]: Also, add steps to 'Add permission' (Delegated => User_impersonation)

File: 07-On demand scan as an API/README.md [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+[Back to top…](README.md#On-this-page)
+
+<!-- TBD: This token is not access subs, but to access AzTS REST API endpoints -->
+
+## Step 2: Configure permissions for WebAPI app registration
+1. Go to Azure Portal.
+2. Go to **App Registration**.
+3. Select your WebAPI App Registration.
+4. Go to **API Permissions**.
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:87]: You can add screenshots. Also, this step is to be performed by the admin.

File: 07-On demand scan as an API/README.md [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+5. Select **Add a permission**.
+6. Go to **APIs my organization uses**.
+7. Search your WebAPI client id and select.
+8. Select **Delegated permissions**.
+9. Select permissions.
+10. **Add permissions**.
+
+
+## Step 3: Get administrator consent for WebAPI app registration
+If 'User consent' is restricted to the WebAPI, then WebAPI must have 'Admin consent' granted to expose the APIs.
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:98]: This statement is incorrect.

If the app consent policies of your Active Directory does not allow user consent, in that case, the consent to access this AzTS API endpoint has to be granted to the Admin of the Active Directory.

Also, this step is for AzTS Admin.

File: 07-On demand scan as an API/README.md [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+2. Go to **App Registration**.
+3. Select WebAPI App Registration.
+4. Go to **Add a client application**.
+5. Add client id of request client app.
+6. Enable the check box **Authorized scopes**.
+7. Add application.
+8. Copy scope from 'Scopes'.
+
+
+> In order to generate the token for APIs, you have to get access for the client application from WebAPI owner.
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:134]: The section **Steps for 'WebAPI Owner' to grant the permission for requested Client app:** should come after this note.

File: 07-On demand scan as an API/README.md [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+Command to generate the token:
+``` PowerShell
+
+$token = Get-MsalToken -TenantId '<tenant-id>' -ClientId '<client-app-id>' -RedirectUri 'https://localhost' -Scopes '<WebAPI-scope>'
+
+```
+
+[Back to top…](README.md#On-this-page)
+
+### Required roles:
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:164]: This can also be added to pre-requisite

File: 07-On demand scan as an API/README.md [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+- CoAdministrator
+- AccountAdministrator
+- Security Reader
+- Security Admin
+> **Note:** You need to run RBAC after granting the permission to SPN over subscription.
+
+## API Operation Groups
+|Operation group|Description|
+|----|----|
+| [Request for scan](README.md#Request-for-scan) |Take subscription(s) for adhoc scan.|
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:180]: Request on-demand scan for a group of subscriptions.

File: 07-On demand scan as an API/README.md [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+| [Get control scan result](README.md#Get-control-scan-result) |Get control scan result for specified subscription.|
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:181]: Get control scan result of a subscription.

File: 07-On demand scan as an API/README.md [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+
+
+## Request for scan
+Take list of subscription id(s) for adhoc scan.
+
+## Description
+To scan a subscription, you can pass list of subscription id(s). This API will return metadata about the status of subscription including 'Scan Request Id'. This 'Scan Request Id' can be further use to [get latest control scan result](README.md#Get-control-scan-result).
+
```
Comment (SINIKI) [inline 07-On demand scan as an API/README.md:189]: can be further use**d** to

File: 07-On demand scan as an API/README.md [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+- [Overview](README.md#overview)
+- [Create client app registration](README.md#Create-client-app-registration)
+- [Generate user authentication token to access subscriptions](README.md#Generate-user-authentication-token-to-access-subscriptions)
+- [API to scan a subscription](README.md#API-to-scan-a-subscription)
+- [API to get control scan result](README.md#API-to-get-control-scan-result)
+- [Feedback](README.md#Feedback)
+
+-----------------------------------------------------------------
+## Overview 
+The Azure Tenant Security Solution (AzTS) exposes APIs for users to allow on demand scan for a subscription and get control scan result using APIs.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done.

File: 07-On demand scan as an API/README.md [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+Follow below steps to create client application:
+1. Go to the Azure portal to register your application.
+
+2. Search for and select App registrations.
+
+3. Select New registration.
+
+4. When the Register an application page appears, enter your application's registration information:
+
+    - In the Name section, enter a meaningful application name that will be displayed to users of the app, such as client-app.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Removed 'client-app'.

File: 07-On demand scan as an API/README.md (hunk 14 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:180]: Done

File: 07-On demand scan as an API/README.md (hunk 15 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:181]: Done

File: 07-On demand scan as an API/README.md (hunk 16 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:189]: Done

File: 07-On demand scan as an API/README.md (hunk 13 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:164]: Added 'required roles' into pre-requisite.

File: 07-On demand scan as an API/README.md (hunk 12 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:134]: Done

File: 07-On demand scan as an API/README.md (hunk 10 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:87]: Screenshot added

File: 07-On demand scan as an API/README.md [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+
+## Prerequisites (For AzTS admin)
+Admin has to enable the flag **FeatureManagement__OnDemandScanAPI** in order to provide access over API endpoints.
+How to enable 'OnDemandScanAPI' from Azure Portal:
+1. Go to Azure Portal.
+2. Go to **Resource Groups**.
+3. Select your Resource Group where you have configured AzTS set up.
+4. Select the App Service for API 'AzSK-ATS-API-xxxxx'.
+5. Go to **Configuration**.
+6. Set **FeatureManagement__OnDemandScanAPI** as true.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:42]: Updated.

File: 07-On demand scan as an API/README.md [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+4. Generate authentication token to access subscriptions.
+5. Use the access token to call API endpoints.
+
+## Prerequisites (For AzTS admin)
+Admin has to enable the flag **FeatureManagement__OnDemandScanAPI** in order to provide access over API endpoints.
+How to enable 'OnDemandScanAPI' from Azure Portal:
+1. Go to Azure Portal.
+2. Go to **Resource Groups**.
+3. Select your Resource Group where you have configured AzTS set up.
+4. Select the App Service for API 'AzSK-ATS-API-xxxxx'.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:40]: Yes, it's 'AzSK-ATS-WebAPI-xxxxx'.
Updated.

File: 07-On demand scan as an API/README.md [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+
+This document will walk you through:
+1. Register an application in Azure AD to represent a client application.
+2. Configure permissions for WebAPI app registration.
+3. Get administrator consent for WebAPI app registration.
+4. Generate authentication token to access subscriptions.
+5. Use the access token to call API endpoints.
+
+## Prerequisites (For AzTS admin)
+Admin has to enable the flag **FeatureManagement__OnDemandScanAPI** in order to provide access over API endpoints.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:35]: Updated.

File: 07-On demand scan as an API/README.md (hunk 9 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Added.

Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done

Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done

File: 07-On demand scan as an API/README.md (hunk 8 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Steps added to get client id using WebAPI config.

File: 07-On demand scan as an API/README.md [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+> </br>
+
+Commands to generate the token:
+``` PowerShell
+
+$ClientSecret = '<client-secret>' | ConvertTo-SecureString -AsPlainText -Force
+
+$token = Get-MsalToken -TenantId '<tenant-id>' -ClientId '<client-id>' -ClientSecret $ClientSecret -Scopes "<WebAPI-scope>/.default"
+
+$header = "Bearer " + $token.AccessToken
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done.

File: 07-On demand scan as an API/README.md [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+- Security Reader
+- Security Admin
+> **Note:** You need to run RBAC after granting the permission to SPN over subscription.
+
+### Using client credential flow
+Client crediential flow uses the client credentials (client id and client secret) to generate the token. Token will be generated against specified SPN (Service Principal Name) and **SPN must have access over the subscription** to scan or to get the control scan result.
+
+> In order to generate the token for APIs, you have to get access for the client application from WebAPI owner.
+> 1. Send the client id to WebAPI owner to request access for client application.
+> 2. WebAPI owner will grant the access and share the scope.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:136]: Done.

File: 07-On demand scan as an API/README.md [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+- AccountAdministrator
+- Security Reader
+- Security Admin
+> **Note:** You need to run RBAC after granting the permission to SPN over subscription.
+
+### Using client credential flow
+Client crediential flow uses the client credentials (client id and client secret) to generate the token. Token will be generated against specified SPN (Service Principal Name) and **SPN must have access over the subscription** to scan or to get the control scan result.
+
+> In order to generate the token for APIs, you have to get access for the client application from WebAPI owner.
+> 1. Send the client id to WebAPI owner to request access for client application.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:135]: Done.

File: 07-On demand scan as an API/README.md [hunk 25]
```diff
@@ -0,0 +1,252 @@
...
+There are two ways to generate access tokens:
+- Using client credential flow
+- Using user authentication code flow
+
+#### Required Az Module
+``` PowerShell
+Install-Module -Name MSAL.PS -AllowClobber -Scope CurrentUser -repository PSGallery
+```
+
+### Required roles:
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done

File: 07-On demand scan as an API/README.md (hunk 7 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done

File: 07-On demand scan as an API/README.md (hunk 6 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Added this point in appropriate API section

File: 07-On demand scan as an API/README.md (hunk 5 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done

File: 07-On demand scan as an API/README.md (hunk 4 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Added as part of pre-requisites.

File: 07-On demand scan as an API/README.md (hunk 3 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:249]: Added description + important points as part of data contract.

File: 07-On demand scan as an API/README.md (hunk 2 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Added description + data contract as part of 'Notes'.

File: 07-On demand scan as an API/README.md (hunk 1 above)
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done

File: 07-On demand scan as an API/README.md [hunk 26]
```diff
@@ -0,0 +1,203 @@
...
+$token.AccessToken | ConvertTo-Json | Out-File '<token file path>\token.json'
+
+```
+
+[Back to top…](README.md#On-this-page)
+
+### Using user authentication code flow
+User authentication code flow uses user's crediential to generate the token. User must have access over the subscription  to scan or to get the control scan result.
+
+> In this flow, Client app registration needs "Admin consent" to access the APIs.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Updated statement.

File: 07-On demand scan as an API/README.md [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+$token = Get-MsalToken -TenantId '<tenant-id>' -ClientId '<client-id>' -ClientSecret $ClientSecret -Scopes "<WebAPI-scope>/.default"
+
+$token.AccessToken | ConvertTo-Json | Out-File '<token file path>\token.json'
+
+```
+
+[Back to top…](README.md#On-this-page)
+
+### Using user authentication code flow
+User authentication code flow uses user's crediential to generate the token. User must have access over the subscription  to scan or to get the control scan result.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:152]: Added required role list.

File: 07-On demand scan as an API/README.md [hunk 28]
```diff
@@ -0,0 +1,203 @@
...
+``` PowerShell
+Install-Module -Name MSAL.PS -AllowClobber -Scope CurrentUser -repository PSGallery
+```
+
+### Using client credential flow
+Client crediential flow uses the client credentials(client id and client secret) to generate the token. Token will be generating against specified SPN(Service Principal Name) and **SPN must have access over the subscription** to scan or to get the control scan result.
+
+> In order to generate the token for APIs, you have to get access for the client application from WebAPI owner.
+> 1. Send the client id to WebAPI owner to grant access for client application.
+> 2. WebAPI owner will grant the access and share the scope.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:136]: Added steps to get scope.

File: 07-On demand scan as an API/README.md [hunk 29]
```diff
@@ -0,0 +1,203 @@
...
+#### Required Az Module
+``` PowerShell
+Install-Module -Name MSAL.PS -AllowClobber -Scope CurrentUser -repository PSGallery
+```
+
+### Using client credential flow
+Client crediential flow uses the client credentials(client id and client secret) to generate the token. Token will be generating against specified SPN(Service Principal Name) and **SPN must have access over the subscription** to scan or to get the control scan result.
+
+> In order to generate the token for APIs, you have to get access for the client application from WebAPI owner.
+> 1. Send the client id to WebAPI owner to grant access for client application.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done

File: 07-On demand scan as an API/README.md [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+- Using client credential flow
+- Using user authentication code flow
+
+#### Required Az Module
+``` PowerShell
+Install-Module -Name MSAL.PS -AllowClobber -Scope CurrentUser -repository PSGallery
+```
+
+### Using client credential flow
+Client crediential flow uses the client credentials(client id and client secret) to generate the token. Token will be generating against specified SPN(Service Principal Name) and **SPN must have access over the subscription** to scan or to get the control scan result.
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done

[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Space added

File: 07-On demand scan as an API/README.md [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+    - Under Add a client secret, provide a Description. Choose when the key should expire, and select Add.
+
+9. When the secret is created, note the key value for use in a subsequent step.
+
+</br>
+
+
+[Back to top…](README.md#On-this-page)
+
+## Generate user authentication token to access subscriptions
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Updated

File: 07-On demand scan as an API/README.md [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+Before generating the token or making an API call, client app registration needs to be created.
+</br>
+Follow below steps to create client application:
+1. Go to the Azure portal to register your application.
+
+2. Search for and select App registrations.
+
+3. Select New registration.
+
+4. When the Register an application page appears, enter your application's registration information:
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md]: Done

File: 07-On demand scan as an API/README.md [hunk 33]
```diff
@@ -0,0 +1,312 @@
...
+- [Generate user authentication token to access subscriptions](README.md#Generate-user-authentication-token-to-access-subscriptions)
+- [API to scan a subscription](README.md#API-to-scan-a-subscription)
+- [API to get control scan result](README.md#API-to-get-control-scan-result)
+- [Feedback](README.md#Feedback)
+
+-----------------------------------------------------------------
+## Overview 
+The Azure Tenant Security Solution (AzTS) provides APIs for users to allow on demand scan for a subscription and get control scan result.
+
+This document will walk you through:
```
[the remark this answers is not in the document]
Comment (Aditi-sha) [reply 07-On demand scan as an API/README.md:23]: Updated statement

**Review by TarunKrShukla — APPROVED** [review]
Please do add task to track future enhancements in this doc.

**Pull request merged**